{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scientific Programming Exam, Monday 05, February 2018\n",
    "\n",
    "**Scientific Programming Labs, Part A + Part B, Quantitative Computational Biology Master, CIBIO**\n",
    "\n",
    "## [Download exercises + solution](../../_static/sciprolab2-2018-02-05-exam.zip) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Introduction\n",
    "\n",
    "\n",
    "**Part A** is by Luca Bianco: \n",
    "\n",
    "- A.1 computeGeneStats\n",
    "- A.2 printSequence\n",
    "\n",
    "**Part B.1**  is more about theory with dynamic programming, and will be assessed by Alberto Montresor:\n",
    "\n",
    "- B.1 subsetsum\n",
    "\n",
    "**Part B.2**  is a LinkedQueue exercise by David Leoni \n",
    "\n",
    "- B.2.1 enqn \n",
    "- B.2.2 deqn\n",
    "\n",
    "\n",
    "### Allowed material\n",
    "\n",
    "There won't be any internet access. You will only be able to access:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "raw_mimetype": "text/html"
   },
   "source": [
    "\n",
    "<ul>\n",
    "<li><a href=\"index.html\" target=\"_blank\">Scientific Programming Algolab worksheets</a></li>\n",
    "<li><a href=\"http://cricca.disi.unitn.it/montresor/teaching/scientific-programming/slides/index.html\" target=\"_blank\">Alberto Montresor slides</a></li>\n",
    "<li><a href=\" /usr/local/sciprolab1/\" target=\"_blank\">Luca Bianco docs</a></li>\n",
    "<li>Python 3 documentation :&nbsp;&nbsp; <a href=\"../python-docs/html/index.html\" target=\"_blank\">html</a>\n",
    "&nbsp;&nbsp;<a href=\"../python-docs/pdf\" target=\"_blank\">pdf</a>\n",
    "<br/> In particular, <a href=\"../python-docs/html/library/unittest.html\" target=\"_blank\">Unittest docs</a>\n",
    "</li>\n",
    "<li>The course book <i>Problem Solving with Algorithms and Data Structures using Python</i> &nbsp;&nbsp;<a href=\"../pythonds/index.html\" target=\"_blank\">html</a> &nbsp;&nbsp;&nbsp;<a href=\"../pythonds/ProblemSolvingwithAlgorithmsandDataStructures.pdf\" target=\"_blank\">pdf</a>\n",
    "</li>\n",
    "</ul>\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "raw_mimetype": "text/latex"
   },
   "source": [
    "\\begin{itemize}  \n",
    "\\item Scientific Programming Algolab worksheets\n",
    "\\item Alberto Montresor slides\n",
    "\\item Luca Bianco docs\n",
    "\\item Python 3 documentation (in particular, see unittest)\n",
    "\\item The course book \"Problem Solving with Algorithms and Data Structures using Python\"\n",
    "\\end{itemize}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Grading\n",
    "\n",
    "\n",
    "* **Lab grade:** The grade of this lab part will range from 0 to 30. \n",
    "* **Correct implementations**: Correct implementations with the required complexity grant you full grade.\n",
    "* **Partial implementations**: Partial implementations _might_ still give you a few points. If you just can't solve an exercise,  try to solve it at least for some subcase (i.e. array of fixed size 2) commenting why you did so. \n",
    "* **Bonus point**: One bonus point can be earned by writing stylish code. You got style if you:\n",
    "\n",
    "    - do not infringe the [Commandments](../../intro.ipynb#Commandments)\n",
    "    - write [pythonic code](http://docs.python-guide.org/en/latest/writing/style)\n",
    "    - avoid convoluted code like i.e.\n",
    "\n",
    "        ```\n",
    "        if x > 5:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        ```\n",
    "\n",
    "        when you could write just\n",
    "\n",
    "        ```\n",
    "        return x > 5\n",
    "        ```       \n",
    "\n",
    "### Valid code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING**: MAKE SURE ALL EXERCISE FILES AT LEAST COMPILE !!! \n",
    "10 MINS BEFORE THE END OF THE EXAM I WILL ASK YOU TO DO A FINAL CLEAN UP OF THE CODE \n",
    "</div>\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING**: _ONLY_ IMPLEMENTATIONS OF THE PROVIDED FUNCTION SIGNATURES WILL BE EVALUATED !!!!!!!!! <br/>    \n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "For example, if you are given to implement: \n",
    "\n",
    "```python\n",
    "    def f(x):\n",
    "        raise Exception(\"TODO implement me\")\n",
    "```\n",
    "\n",
    "and you ship this code:\n",
    "\n",
    "``` python\n",
    "def my_f(x):\n",
    "    # a super fast, correct and stylish implementation\n",
    "\n",
    "def f(x):\n",
    "    raise Exception(\"TODO implement me\")\n",
    "```\n",
    "\n",
    "We will assess only the latter one `f(x)`, and conclude it doesn't work at all :P   !!!!!!!\n",
    "\n",
    "**Helper functions**\n",
    "\n",
    "Still, you are allowed to define any extra helper function you might need. If your `f(x)` implementation calls some other function you defined like `my_f` here, it is ok:\n",
    "\n",
    "```python\n",
    "# Not called by f, will get ignored:\n",
    "def my_g(x):\n",
    "    # bla\n",
    "\n",
    "# Called by f, will be graded:\n",
    "def my_f(y,z):\n",
    "    # bla\n",
    "\n",
    "def f(x):\n",
    "    my_f(x,5)\n",
    "    \n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "### How to edit and run\n",
    "\n",
    "To edit the files, you can use any editor of your choice: \n",
    "* **Editra editor** is easy to use, you can find it under _Applications->Programming->Editra_.\n",
    "* Others could be _GEdit_ (simpler), or _PyCharm_ (more complex).\n",
    "\n",
    "To run the tests, use **the Terminal** which can be found in _Accessories -> Terminal_ \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**IMPORTANT**: Pay close attention to the comments of the functions.\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING**: _DON'T_ modify function signatures! Just provide the implementation.\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING**: <i>DON'T</i> change the existing test methods, just add new ones !!!  You can add as many as you want.\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING**: _DON'T_ create other files. If you still do it, they won't be evaluated.\n",
    "</div>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Debugging \n",
    "\n",
    "If you need to print some debugging information, you are allowed to put extra <code>print</code>\n",
    "    statements in the function bodies."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING:** even if print statements are allowed, be careful with prints that might\n",
    "     break your function! \n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example, avoid stuff like this: \n",
    "\n",
    "```python\n",
    "x = 0\n",
    "print(1/x)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What to do\n",
    "\n",
    "1) Download  [sciprolab2-2018-02-05-exam.zip](../../_static/sciprolab2-2018-02-05-exam.zip) and extract it *on your desktop*. Folder content should be something like this:\n",
    "\n",
    "```\n",
    "sciprolab2-2018-02-05\n",
    "    |- FIRSTNAME-LASTNAME-ID\n",
    "        |- exerciseA1.py\n",
    "        |- exerciseB1.py\n",
    "        |- exerciseB2.py    \n",
    "\n",
    "```\n",
    "\n",
    "2) Rename `FIRSTNAME-LASTNAME-ID` folder: put your name, lastname an id number, like `john-doe-432432`\n",
    "\n",
    "From now on, you will be editing the files in that folder. At the end of the exam, that is what will be evaluated.\n",
    "\n",
    "3) Edit the files following the instructions in this worksheet for each exercise. Every exercise should take max 25 mins. If it takes longer, leave it and try another exercise.\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part A  \n",
    "\n",
    "by Luca Bianco. \n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "Please write only in exerciseA.py, don't create other files!<br/>\n",
    "\n",
    "**IMPORTANT**: Add your name and ID (matricola) on top of the .py file!\n",
    "</div>\n",
    "\n",
    "###  The Problem\n",
    "\n",
    "The `.tsv` file `gene_models.tsv` is a compact representation of the gene models on the corresponding genome (the genome sequence is present in the other file, `sequences.fasta`, in fasta format). The first few lines are reported below:\n",
    "\n",
    "```\n",
    "Chr     feature start   end     ID\n",
    "Chr05   gene    50988   51101   gene:MD05G1000100\n",
    "Chr05   ncRNA   50988   51101   ncRNA:MD05G1000100\n",
    "Chr05   gene    83210   83329   gene:MD05G1000200\n",
    "Chr05   ncRNA   83210   83329   ncRNA:MD05G1000200\n",
    "Chr05   gene    87650   91333   gene:MD05G1000300\n",
    "Chr05   mRNA    87650   91333   mRNA:MD05G1000300\n",
    "Chr05   CDS     87650   87727   CDS:MD05G1000300.8\n",
    "```\n",
    "\n",
    "as the header (first line) describes, the first column is the chromosome containing the feature (that can be gene, `ncRNA`, `exon`, `CDS`, ...), the third and fourth columns are the start and end position of the feature on the chromosome and the fifth row is the identifier of the feature `ID`.\n",
    "The file `sequences.fasta` stores sequence information and the corresponding sequencing quality. A mock entry is the following:\n",
    "\n",
    "```\n",
    ">Chr01 \n",
    "AGGCCTAGGTCTTCCAGAGTCGCTTTTTCCAGCTCCAGACCGATCTCTTCAG\n",
    "AGGCCAATCGCCAGTTTACCACATACACCCAGACCGATCTCTTCAG\n",
    "```\n",
    "\n",
    "where the first line is the identifier of the read and starts with a `>`. The sequence follows the line with the identifier and can be on multiple lines.\n",
    "\n",
    "\n",
    "\n",
    "### A.1) computeGeneStats(filename) \n",
    "\n",
    "Implement function `computeGeneStats(filename)`: gets the filename of a `.tsv` file as explained above, stores its content in a suitable data structure of your choice (hint: pandas might help here), counts (and prints) the number of features of type gene, printing the average gene length in the whole file. The function should also plot a bar plot of the number of genes per chromosome.\n",
    "\n",
    "**Note: The function should return the data structure containing all the data.**\n",
    "\n",
    "Calling:\n",
    "\n",
    "```\n",
    "fn = \"gene_models.tsv\"\n",
    "seqFile = \"sequences.fasta\"\n",
    "GenesDF = computeGeneStats(fn)\n",
    "```\n",
    "\n",
    "should give:\n",
    "\n",
    "```\n",
    "29136 genes present\n",
    "Avg gene length: 2723.71\n",
    "```\n",
    "\n",
    "![](compute-gene-stats.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A.2) printSequence(geneInfo, geneID, sequenceFile)\n",
    "\n",
    "Implement `printSequence(geneInfo, geneID, sequenceFile)`: gets the `geneInfo` data structure created by `computeGeneStats`, a gene identifier `geneID` and the filename of a fasta formatted file `sequenceFile` and if `geneID` is present in `geneInfo` it prints its length. The function must also retrieve the sequence of `geneID` (chromosome, start and end position are saved in `geneInfo`) if it is present in the fasta file `sequenceFile`. If the chromosome is not in the `sequenceFile` or `geneID` is not in `geneInfo` the function should display a corresponding message (see examples below).\n",
    "\n",
    "**Hint: you can use biophtyon to read a fasta file.** \n",
    "\n",
    "**Examples**\n",
    "\n",
    "For example, assuming an entry:\n",
    "```\n",
    "ChrX   gene   1   10   gene:geneid1\n",
    "```\n",
    "in `geneInfo` and a sequence:\n",
    "\n",
    "```\n",
    ">ChrX\n",
    "GATTACATAACACACTACA\n",
    "```\n",
    "\n",
    "in `sequenceFile`, calling `printSequence(geneInfo,\"geneid1\",sequenceFile)` should print:\n",
    "\n",
    "```\n",
    "Gene geneid1 is in ChrX and has length 9\n",
    ">geneid1\n",
    "GATTACATAA\n",
    "```\n",
    "\n",
    "Given `GenesDF` as returned by `computeGeneStats`, calling:\n",
    "\n",
    "```\n",
    "printSequence(GenesDF,\"MD05G1027300\",seqFile)\n",
    "print(\"\")\n",
    "printSequence(GenesDF,\"MD03G1000400\",seqFile)\n",
    "print(\"\")\n",
    "printSequence(GenesDF,\"MD08G1000100\",seqFile)\n",
    "print(\"\")\n",
    "printSequence(GenesDF,\"MD08G100019191\",seqFile)\n",
    "```\n",
    "\n",
    "should give:\n",
    "\n",
    "```\n",
    "\n",
    "Gene MD05G1027300 is in Chr05 and has length 567\n",
    ">MD05G1027300\n",
    "GAAACAGAGATAAAAAGTATCTAAATTCACGCAAATTAACAATTTTCCCATTTCATAGAATCCAAAAGCAGATATATAATTCACCTTTCCCAGATTAAAGCGTGAAACAACCACACCCTTCATCAATCCAAAACCAGGGAAAGAAATGTGTTTTCAAAACATGATAGCATAAATTAATCAAATTACCTTAGAATAGGGGGTCTCCAAGTCGGTGGAGAAGTGTAAGTCGCAGGTGGAGGAGCGGGACGAGGTCTTGGCCGGAGAGTGGTTGAGATCAGAGACGGAGCGGCGGAGAGTTGGCGGTGGAGGAAGAGGAGGCAGGACCCGTGCCGGGGAATGGGACGGACTCTTGGGAAAATGGTGGCAGTGAGGAGAGTTGGGAAGAGTCATATTGGGCAAAGAACCGCCACAGCGAGAGGCAGGAACAAAGTTTTCCAACGGGTTATTGGTGCTGGACGCAGACGGAGGGAGAGAAAATTCCTGTTCCTGACTGTTCGATGCAAGCTGCTAGGGGTGGTGGTGGTGGTGGTGGTTCATGGTGGGCGGCGGAGAAATTGGGTCTTGGGAT\n",
    "\n",
    "Gene MD03G1000400 is in Chr03 and has length 188\n",
    ">MD03G1000400\n",
    "CGTATTATAGGAATTCTTCTTCCATTGAAAGGATGGATTGAAAGGATGGATGGCATGTTGTCTATACCTCTCGTCATTAGGATGTACAATAATGGCAGTATCACCGAGCATAGTTTCCACTCTAGTGGTGGCCACCATAATCTCACCAAAGTCTTCCTCTAGAGTATAAGCAAATGAAGTTAACATGCC\n",
    "\n",
    "Gene MD08G1000100 is in Chr08 and has length 5383\n",
    "Chr08 is not present in the sequence file.\n",
    "\n",
    "GeneID MD08G100019191 is not present in the input file.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Part B \n",
    "\n",
    "### B.1) `SubsetSumTest` \n",
    "\n",
    "By Alberto Montresor. \n",
    "\n",
    "#### The problem \n",
    "\n",
    "Let `A` be a list of non-negative (possibly repeated) integers, and `S` a target integer value. Your task is to decide whether there exists a subset of the integers in `A` whose sum is equal to the target value `S`. You need to write a function `subsetsum(A,S)` that returns `True` if there is such subset, `False` otherwise.\n",
    "\n",
    "Suggestion: the problem is similar to the knapsack problem. You should compute a table entry for each pair `(i,s)`, where `DP[i][s]` should contain `True` if and only if it is possible to obtain `s` by summing a subset of the first `i` items.\n",
    "\n",
    "You should express the problem based on two possibilities:\n",
    "- if you take the last element `A[i]`, the new target is `s-A[i]` and there are `i-1` elements left\n",
    "- if you don't take the last element `A[i]`, the target remain `s` and there are `i-1` elements left. If either of these cases is `True`, you should return `True`.\n",
    "\n",
    "Pay attention to the base cases.\n",
    "\n",
    "Example: if `A=[2,5,4,3,12]` and `S=9`, the answer is `True`; if `A=[2,5,4,3,12]` and `S=25`, the answer is `False`. \n",
    "\n",
    "For plenty of other examples, see the test cases.\n",
    "\n",
    "**Implementation**:\n",
    "\n",
    "Start editing the file `exerciseB1.py` which contains the function:\n",
    "\n",
    "\n",
    "```python\n",
    "\n",
    "def subsetsum(A,S):\n",
    "    \"\"\"\n",
    "    Let A be a list of non-negative integers, and S a target integer value.\n",
    "    Return True if there exists a subset  of the integers in A whose sum is\n",
    "    equal to the target value S. Otherwise, return False.\n",
    "\n",
    "    NOTE:\n",
    "      - if S = 0, always return True, even with empty array\n",
    "      - numbers in A may be repeated\n",
    "    \"\"\"\n",
    "```\n",
    "\n",
    "**Testing**: `python3 -m unittest exerciseB1_test`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B.2) LinkedQueue \n",
    "\n",
    "You are given a queue implemented as a LinkedList, with usual `_head` pointer plus additional `_tail` pointer and `_size` counter\n",
    "\n",
    "- Data in enqueued at he right, in the tail\n",
    "- Data is dequeued at the left, removing it from the head\n",
    "\n",
    "Example, where the arrows represent _next pointers:\n",
    "\n",
    "```\n",
    "_head                        _tail\n",
    "    a -> b -> c -> d -> e -> f\n",
    "```\n",
    "\n",
    "In this exercise you will implement the methods `enqn(lst)` and `deqn(n)` which respectively enqueue a python list of n elements and dequeue n elements, returning python a list of them.\n",
    "\n",
    "Here we show an example usage, see to next points for detailed instructions.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "```\n",
    "        q = LinkedQueue()\n",
    "\n",
    "        q.enqn(['a','b','c'])\n",
    "\n",
    "        Return nothing, queue becomes:\n",
    "        \n",
    "          _head         _tail\n",
    "              a -> b -> c\n",
    "\n",
    "        q.enqn(['d'])\n",
    "\n",
    "        Return nothing, queue becomes:\n",
    "        \n",
    "          _head              _tail\n",
    "              a -> b -> c -> d\n",
    "\n",
    "\n",
    "        q.enqn(['e','f'])\n",
    "\n",
    "        Return nothing, queue becomes:\n",
    "        \n",
    "          _head                        _tail\n",
    "              a -> b -> c -> d -> e -> f\n",
    "\n",
    "        q.deqn(3)\n",
    "\n",
    "        Returns ['d', 'e', 'f'] and queue becomes:\n",
    "\n",
    "          _head         _tail\n",
    "              a -> b -> c\n",
    "\n",
    "        q.deqn(1)\n",
    "\n",
    "        Returns ['c'] and queue becomes:\n",
    "\n",
    "          _head    _tail\n",
    "              a -> b\n",
    "\n",
    "        q.deqn(5)\n",
    "\n",
    "        raises LookupError as there aren't enough elements to remove\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### B.2.1) `enqn`\n",
    "\n",
    "Implement the method `enqn`:\n",
    "\n",
    "```python\n",
    "    def enqn(self, lst):\n",
    "        \"\"\" Enqueues provided list of elements at the tail of the queue\n",
    "\n",
    "            - Required complexity: O(len(lst))\n",
    "            - NOTE: remember to update the _size and _tail\n",
    "\n",
    "            Example: supposing arrows represent _next pointers:\n",
    "\n",
    "          _head         _tail\n",
    "              a -> b -> c\n",
    "\n",
    "            Calling\n",
    "\n",
    "            q.enqn(['d', 'e', 'f', 'g'])\n",
    "\n",
    "            will produce the queue:\n",
    "\n",
    "          _head                             _tail\n",
    "              a -> b -> c -> d -> e -> f -> g\n",
    "\n",
    "```\n",
    "\n",
    "**Testing**: `python3 -m unittest exerciseB2_test.EnqnTest`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### B.2.2) `deqn`\n",
    "\n",
    "\n",
    "Implement the method `deqn`:\n",
    "\n",
    "```python\n",
    "\n",
    "    def deqn(self, n):\n",
    "        \"\"\" Removes n elements from the head, and return them as a Python list,\n",
    "            where the first element that was enqueued will appear at the *beginning* of\n",
    "            the returned Python list.\n",
    "\n",
    "            - if n is greater than the size of the queue, raises a LookupError.\n",
    "            - required complexity: O(n)\n",
    "\n",
    "            NOTE 1: return a list of the *DATA* in the nodes, *NOT* the nodes themselves\n",
    "            NOTE 2: DO NOT try to convert the whole queue to a Python list for playing with splices.\n",
    "            NOTE 3: remember to update _size, _head and _tail when needed.\n",
    "\n",
    "\n",
    "            For example, supposing arrows represent _next pointers:\n",
    "\n",
    "\n",
    "          _head                             _tail\n",
    "              a -> b -> c -> d -> e -> f -> g\n",
    "\n",
    "            q.deqn(3) will return the Python list ['a', 'b', 'c']\n",
    "\n",
    "            After the call, the queue will be like this:\n",
    "\n",
    "          _head              _tail\n",
    "              d -> e -> f -> g\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "```\n",
    "\n",
    "**Testing**: `python3 -m unittest exerciseB2_test.DeqnTest`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
